### PR TITLE
Added cell value formatter

### LINKED ISSUE #1209: Modus Data Table - Add support for a value formatter function
### Prerequisites

- [X] I have [searched](https://github.com/trimble-oss/modus-web-components/issues?q=is%3Aissue) for duplicate or closed feature requests
- [X] I have read the [contributing guidelines](https://github.com/trimble-oss/modus-web-components/blob/main/CONTRIBUTING.md)

### Proposal

To achieve something like https://www.ag-grid.com/javascript-data-grid/value-formatters/ for users to write their own formatting logic for cell content. This way we can achieve currency and date formats.

### Motivation and context

https://www.ag-grid.com/javascript-data-grid/value-formatters/

### Which npm package?

@trimble-oss/modus-web-components

### Priority

High

### What is the product name the feature is requested for?

prism

### What is the team name the product belongs to?

e-builder

### What is the division name the team belongs to?

viewpoint

### PR DESCRIPTION
**Pull request by Ankur-Koul**
## Description

Added cell renderer 

Fixes #1209 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (including CHANGELOG updates)
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ export interface ModusDataTableColumn {
   accessorKey: string;
   dataType: ModusColumnDataType;
   id?: string;
-  columns?: ModusDataTableColumn[];
+  // columns?: ModusDataTableColumn[]; //To-do
```
Comment (sankarNous) [inline stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts:8]: give some comments for 'TODO'

**Review by msankaran0712 — CHANGES_REQUESTED** [review]
1. Is it possible to disable sort tooltips on all columns when performing the column resize? because it is interrupting with tooltip messages which is not necessary when resizing a column. You can check with @sankarNous for more details on how to do it.
2. It is okay to show footer in only one example. Please make sure there is no scrollbar appearing.

**Review by msankaran0712 — CHANGES_REQUESTED** [review]
Please make these changes

image

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export type Renderable<TProps> = any | FunctionalComponent<TProps>;
+
+export function cellFormatter<TProps extends object>(
+  Comp: Renderable<TProps>,
+  props: TProps
+): unknown | JSX.Element {
+  return !Comp ? null : isComponent<TProps>(Comp) ? <Comp {...props} /> : Comp;
+}
+
+function isComponent<TProps>(
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx:17]: please add some comments for these functions

**Review by msankaran0712 — CHANGES_REQUESTED** [review]
Please keep these types commented out for now, and add a comment explaining why.

image

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-header.tsx [hunk 3]
```diff
@@ -33,7 +34,9 @@ export const ModusDataTableHeader: FunctionalComponent<
       scope="col">
       {header.isPlaceholder ? null : ( // header.isPlaceholder is Required for nested column headers to display empty cell
         <div class={header.column.getCanSort() && 'can-sort'}>
-          <span>{header.column.columnDef.header}</span>
+          <span>
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/parts/modus-data-table-header.tsx:37]: why do we need a cell formatter for header?

File: stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ export interface ModusDataTableColumn {
   accessorKey: string;
   dataType: ModusColumnDataType;
   id?: string;
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts:7]: we should a add cell property here.

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  component: Renderable<TProps>
+): component is FunctionalComponent<TProps> {
+  return isClassComponent(component) || typeof component === 'function';
+}
+
+function isClassComponent<TProps>(component: Renderable<TProps>) {
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx:23]: I think we must rename the functions to "isFunctionalComponent", "isFunction" as we don't exactly look for a class component

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell.tsx [hunk 6]
```diff
@@ -5,33 +5,31 @@ import {
...
     <td
-      key={props.cell.id}
+      key={cell.id}
       class={`
        ${
-         props.cell.column.columnDef[PropertyDataType] ===
+         cell.column.columnDef[PropertyDataType] ===
            ModusColumnDataType.Integer ||
-         props.cell.column.columnDef[PropertyDataType] ===
+         cell.column.columnDef[PropertyDataType] ===
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell.tsx:24]: once we comment out currency data type, we might need to remove the OR clause

**Review by msankaran0712 — CHANGES_REQUESTED** [review]

**Review by msankaran0712 — CHANGES_REQUESTED** [review]
@Ankur-Koul Can you please also add one column showing currency, one for date, one for decimal places ?

examples:

image

image

File: stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts (hunk 1 above)
Comment (Ankur-Koul) [reply stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts]: Added

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-header.tsx (hunk 3 above)
Comment (Ankur-Koul) [reply stencil-workspace/src/components/modus-data-table/parts/modus-data-table-header.tsx]: Removed

File: stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts (hunk 4 above)
Comment (Ankur-Koul) [reply stencil-workspace/src/components/modus-data-table/models/modus-data-table-column.ts:8]: Added

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx (hunk 5 above)
Comment (Ankur-Koul) [reply stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx:29]: Done

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell.tsx (hunk 6 above)
Comment (Ankur-Koul) [reply stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell.tsx]: Removed

File: stencil-workspace/src/components/modus-data-table/enums/modus-column-data-type.ts [hunk 7]
```diff
@@ -1,9 +1,10 @@
 export enum ModusColumnDataType {
   Text = 'text',
   Integer = 'integer',
-  Currency = 'currency',
-  Percent = 'percent',
-  Decimal = 'decimal',
-  Date = 'date',
-  Dropdown = 'dropdown',
+  // To-do, how to support for these datatypes is not decided for now.
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/enums/modus-column-data-type.ts:4]: please remove "TODO" and update it to "to support in the future"

File: stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ */
+export function cellFormatter<TProps extends object>(
+  Comp: Renderable<TProps>,
+  props: TProps
+): unknown | JSX.Element {
+  return !Comp ? null : isFunction<TProps>(Comp) ? <Comp {...props} /> : Comp;
+}
+
+/**
+ * Checks if the item to display is a component or not.
```
Comment (msankaran0712) [inline stencil-workspace/src/components/modus-data-table/parts/modus-data-table-cell-formatter.tsx:21]: I would recommend updating the text to "checks if the input for cell formatter is a functional component or class component"

File: stencil-workspace/storybook/stories/components/modus-data-table/modus-data-table-storybook-docs.mdx [hunk 9]
```diff
@@ -142,6 +143,42 @@ Users can arrange data in a table by sorting it in either ascending or descendin
 
 <Story id="components-data-table--sorting" height={'350px'} />
 
+### Value Formatter
+
+User can take the data of a cell and modify it according to there requirements
```
Comment (msankaran0712) [inline stencil-workspace/storybook/stories/components/modus-data-table/modus-data-table-storybook-docs.mdx:148]: Recommendation for line 148: "User can take the data of a cell and apply any string-based formatting logic."

**Review by msankaran0712 — CHANGES_REQUESTED** [review]
Scrollbars are appearing for value formatter and column resize examples.

**Review by msankaran0712 — APPROVED** [review]

**Pull request merged**